source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+/++
+    It is sometimes necessary to create a function which is an exact copy of
+    another function. Or sometimes it is necessary to introduce a few
+    variations, while carrying all the other aspects. Because of function
+    attributes, parameter storage classes and user-defined attributes, this
+    requires building a string mixin. In addition, the mixed-in code must refer
+    only to local names, if it is to work across module boundaires.
+
+    This module facilitates the creation of such mixins.
+
+ Copyright: Copyright Jean-Louis Leroy 2017-2020
+
+ License:   $(LINK2 http://boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+ Authors:   Jean-Louis Leroy
++/
+
+module bolts.experimental.refraction;
+
+import std.array;
+import std.format;
+import std.meta;
+import std.algorithm.iteration : map;
+import std.range : iota;
+import std.traits;
+
+template ParameterAttribute(alias F, int i, int j)
+{
+  static if (is(typeof(F) P == __parameters)) {
+    alias ParameterAttribute = Alias!(__traits(getAttributes, P[i..i+1])[j]);
+  }
+}
+
+/**
+    Return a Function object that captures all the aspects of Fun, using the
+    value of localSymbol to represent the return and parameter types, and the
+    UDAs.
+
+    Params:
+    fun = a function
+    localSymbol = a string mixin that represents fun in the caller's context
+
+*/
+
+Function refract(alias fun, string localSymbol)()
+  if (is(typeof(fun) == function))
+{
+  Function model = {
+  name: __traits(identifier, fun),
+  localSymbol: localSymbol,
+  returnType: "std.traits.ReturnType!("~localSymbol~")",
+  parameters: refractParameterList!(fun, localSymbol),
+  udas: __traits(getAttributes, fun)
+  .length.iota.map!(
+    formatIndex!("@(__traits(getAttributes, %s)[%%d])".format(localSymbol))).array,
+  attributes: functionAttributes!(fun),
+  static_: __traits(isStaticFunction, fun) && isAggregate!(__traits(parent, fun)),
+  body_: ";",
+  };
+
+  return model;
+}
+
+private enum isScope(alias T) = is(T == module) || is(T == struct)
+  || is(T == class) || is(T == interface);
+
+Function refract(
+  alias Scope, string localSymbol, string name, uint index)()
+  if (isScope!Scope)
+{
+  return refract!(
+    __traits(getOverloads, Scope, name)[index],
```

Comment:
> [getOverloads](https://dlang.org/spec/traits.html#getOverloads) also takes a `bool` at the end which presents template overloads as well. What kind of semantics do you expect when `name` refers to a template?

It breaks my heart but AFAICT D does not currently make it possible to analyze templates beyond querying some very basic properties. I have been struggling to implement templatized open methods for years, without success, because of that. For example, given:

```d
void foo(T)(lazy T t);
```

...there is no way (short of parsing the `stringof`) of extracting the `lazy`.

In the future, we may add a way of making a `FunctionTemplate` from a `Function` but again the starting point would have to be a mere function :-(